src/test/java/edu/unc/lib/boxc/migration/cdm/services/RedirectMappingIndexServiceTest.java

```diff
@@ -0,0 +1,160 @@
+    private static final String DEST_UUID = "7a33f5e6-f0ca-461c-8df0-c76c62198b17";
+
+    @Rule
+    public final TemporaryFolder tmpFolder = new TemporaryFolder();
+
+    private MigrationProject project;
+    private RedirectMappingIndexService indexService;
+    private SipService sipsService;
+    private SipServiceHelper testHelper;
+
+    @Before
+    public void setup() throws Exception {
+        tmpFolder.create();
+        project = MigrationProjectFactory.createMigrationProject(
+                tmpFolder.getRoot().toPath(), PROJECT_NAME, null, "user");
+        testHelper = new SipServiceHelper(project, tmpFolder.newFolder().toPath());
+        sipsService = testHelper.createSipsService();
+
+        indexService = new RedirectMappingIndexService(project);
+        Connection conn = indexService.openDbConnection();
+        Statement statement = conn.createStatement();
+        statement.execute("drop table if exists redirect_mappings");
+        statement.execute("CREATE TABLE redirect_mappings (" +
+                "id INTEGER PRIMARY KEY AUTOINCREMENT, " +
+                "cdm_collection_id varchar(64) NOT NULL, " +
+                "cdm_object_id varchar(64) NOT NULL, " +
+                "boxc_work_id varchar(64) NOT NULL, " +
+                "boxc_file_id varchar(64) DEFAULT NULL, " +
+                "created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP, " +
+                "UNIQUE (cdm_collection_id, cdm_object_id)" +
+                ")");
+        CdmIndexService.closeDbConnection(conn);
+
+        testHelper.indexExportData("export_1.xml");
+        testHelper.generateDefaultDestinationsMapping(DEST_UUID, null);
+        testHelper.populateDescriptions("gilmer_mods1.xml");
+        testHelper.populateSourceFiles("276_182_E.tif", "276_183B_E.tif", "276_203_E.tif");
+
+        sipsService.generateSips(makeOptions());
+    }
+
+    @Test
+    public void indexingDoesNotHappenIfCollectionIsNotSubmitted() {
+        try {
+            indexService.indexMapping();
+        } catch (InvalidProjectStateException e) {
+            assertTrue("Unexpected message: " + e.getMessage(),
+                    e.getMessage().contains("Must submit the collection prior to indexing"));
+        }
+    }
+
+    @Test
+    public void redirectMappingTableIsPopulatedWithRightNumberOfRows() throws Exception {
+        addSipsSubmitted();
+        indexService.indexMapping();
+
+        Connection conn = indexService.openDbConnection();
+        try {
+            Statement stmt = conn.createStatement();
+            ResultSet rs = stmt.executeQuery("select count(*) from redirect_mappings");
+            rs.next();
+            assertEquals("Incorrect number of rows in database", 3, rs.getInt(1));
+        } finally {
+            CdmIndexService.closeDbConnection(conn);
+        }
+    }
+
+    @Test
+    public void  redirectMappingIndexPopulatesColumnsCorrectly() throws Exception {
+        ArrayList<String> row1 = new ArrayList<>();
```

Comment:
Just a java thing, but you typically want to declare variables by their interface rather than their implementation (`ArrayList` is an implementation of the `List` interface).
```suggestion
        List<String> row1 = new ArrayList<>();
```
In java 11 there are other ways to handle this, but we can look at that when we are able to make the jump.

Reply:
Can you explain this a bit more? I don't mind doing it but I don't understand the difference? You declare it as a List with the variable name but then the instantiation is an ArrayList?

Reply:
The intent is to decouple code from details of the implementation, so that you could swap in other implementations of lists later if need be. This stackoverflow thread talks about it:
https://stackoverflow.com/questions/9852831/polymorphism-why-use-list-list-new-arraylist-instead-of-arraylist-list-n

Generally you only declare a variable as the implementation type if you need functionality specific to that implementation. It seems a bit abstract, but it matters more in some places than others, like if you have a data model that you need to implement for two different back ends (maybe one is database driven and the other is from local files), and have code that uses the data model but doesn't care which implementation is used.

It doesn't matter that much here, since ArrayList only has 2-3 methods not available to the List interface, so mainly its a best practice that you'll see all over the place in java code.